File: src/main/java/uk/co/optimisticpanda/jnomon/formatter/ElapsedLineOutputWriter.java
package uk.co.optimisticpanda.jnomon.formatter;

import static java.lang.String.format;
import static java.lang.System.currentTimeMillis;
import static uk.co.optimisticpanda.jnomon.Utils.BORDER;
import static uk.co.optimisticpanda.jnomon.Utils.SECONDS_FORMAT;
import uk.co.optimisticpanda.jnomon.ColourChooser;
import uk.co.optimisticpanda.jnomon.Utils.Colour;

public class ElapsedLineOutputWriter implements EventListener {
    private final ColourChooser colourChooser;
    
    public ElapsedLineOutputWriter(ColourChooser colourChooser) {
        this.colourChooser = colourChooser;
    }
    
    @Override
    public void onLineStart(long processStartTime, long lastStartTime, String line) {
        onUpdate(lastStartTime, lastStartTime);
        System.out.printf("%8s  %s %s\n", "", BORDER, line);
    }
    
    @Override
    public void onUpdate(long processStartTime, long currentStartTime) {
        long elapsed = currentTimeMillis() - currentStartTime;
        String text = format("%8ss %s", SECONDS_FORMAT.format((elapsed) / 1000d), BORDER);
        Colour colourForDuration = colourChooser.colourForDuration(elapsed);
        System.out.print("\033[1A\r" + colourForDuration.colourize(text) + "\n");
    }
    
    @Override
    public void onLineEnd(long processStartTime, long lastStartTime, String lastLine) {
        long elapsed = currentTimeMillis() - lastStartTime;
        System.out.printf("%8ss %s %s\n", SECONDS_FORMAT.format((elapsed) / 1000d), BORDER, lastLine);
    }
}

File: src/main/java/uk/co/optimisticpanda/jnomon/formatter/ElapsedLineOutputWriter.kt
package uk.co.optimisticpanda.jnomon.formatter

import uk.co.optimisticpanda.jnomon.Utils.Colour
import uk.co.optimisticpanda.jnomon.Utils.MOVE_TO_START_OF_PREV_LINE
import uk.co.optimisticpanda.jnomon.Utils.formatLine
import uk.co.optimisticpanda.jnomon.Utils.formatSeconds
import java.time.Duration

class ElapsedLineOutputWriter : EventListener {
  override fun onLineStart(colour: Colour?, sinceProcessStart: Duration, sinceLastStart: Duration, line: String?) {
    onUpdate(colour, sinceProcessStart, sinceLastStart)
    printer().println(formatLine(8, "", line))
  }

  override fun onUpdate(colour: Colour?, sinceProcessStart: Duration, sinceLastStart: Duration) {
    val text: String = formatLine(8, formatSeconds(sinceLastStart), "")
    printer().println(MOVE_TO_START_OF_PREV_LINE + colour!!.colourize(text))
  }

  override fun onLineEnd(sinceProcessStart: Duration, sinceLastStart: Duration, lastLine: String?) {
    printer().println(formatLine(8, formatSeconds(sinceLastStart), lastLine))
  }

  override fun onFinally(sinceProcessStart: Duration, sinceLastStart: Duration?) {
    printer().println(formatLine(8, "Total:", formatSeconds(sinceProcessStart)))
  }
}